### PR TITLE
MB-10977 update existing load tests to use new helpers

### PR DESCRIPTION
**Pull request by felipe-lee**
## Description

Converting existing load tests to follow new patterns/use new helpers. I was documenting it for another team, but it ended up being faster to do it than write down examples and explanations of each conversion.

There were a few more changes I made to base helper code:

* Add ability for the `MilMoveRequestPreparer` to handle the `name` kwarg for requests.
* Break down `check_response` into a few logging functions. This makes it easier for each test to be more specific about what they expect and to test the helper functions. It also enables us to use them for both regular `Responses` (from using `requests` package directly) and `RestResponseContextManager` (the ones we get from `self.rest`) instances.

## Reviewer Notes

I ran the load tests and am getting similar results for the most part, with a few improvements (the expected failures now don't get counted as failures). I think this code is ready for review.

It might be easier to review it if you go commit by commit. Also, if it's hard to review, I'm happy to do a meeting, or I think there's also a few commits at the end (the `check_response` changes) that could even be split into a second pr.

## Setup

You can run the unit tests:

```shell
pytest
```

the prime load tests:

```shell
make load_test_prime
```

the office load tests:

```shell
make load_test_office
```

and the milmove load tests:

```shell
make load_test_milmove
```

Though the last one only really creates a user and logs out their email.

## Code Review Verification Steps

* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-10977) for this change.

### CODE REVIEW

File: locustfiles/prime_workflow.py [hunk 1]
```diff
@@ -3,19 +3,13 @@
...
-    local_port = "9443"
-    domain = MilMoveDomain.PRIME  # the base domain for the host
-
-    wait_time = between(0.25, 9)  # the time period to wait in between tasks (in seconds, accepts decimals and 0)
-    weight = 1
-
-    # For the tasks, we can define all the workflow tasksets we have set up:
     tasks = {PrimeWorkflowTasks: 1, PrimeEndpointWorkflowsTasks: 1}
+    wait_time = between(0.25, 9)
+    weight = 1
```
Comment (duncan-truss) [inline locustfiles/prime_workflow.py:15]: This isn't new to your PR but I just noticed this has a weight when the other User locustfiles don't specify one.  Any idea if we need this?

Comment (felipe-lee) [reply locustfiles/prime_workflow.py:15]: Oh right, I forgot to look into it when I was making the conversions. I'm not sure, I'll check.

Comment (felipe-lee) [reply locustfiles/prime_workflow.py:15]: Ah apparently it does have a use, but in this case we're using the default so we could get rid of it: https://docs.locust.io/en/stable/api.html#locust.User.weight

Comment (duncan-truss) [reply locustfiles/prime_workflow.py:15]: Thanks, yea it didn't seem different than the default.

**Comment by duncan-truss** [comment]
Is `host` a required flag?  I feel like before if it wasn't specified we defaulted to local but I get this error if I omit it

```
➜  milmove_load_testing git:(MB-10977-fl-update-existing-load-tests-to-use-new-helpers) ✗ locust -f locustfiles/prime.py
[2022-01-12 14:13:39,337] Duncans-MacBook-Pro.local/INFO/locust.main: Starting web interface at http://0.0.0.0:8089 (accepting connections from all network interfaces)
[2022-01-12 14:13:39,351] Duncans-MacBook-Pro.local/ERROR/root: Uncaught exception in event handler: 
Traceback (most recent call last):
  File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/event.py", line 40, in fire
    handler(**kwargs)
  File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 48, in on_init
    raise err
  File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 42, in on_init
    milmove_env = MilMoveEnv(value=environment.host)
  File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 384, in __call__
    return cls.__new__(cls, value)
  File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 702, in __new__
    raise ve_exc
ValueError: None is not a valid MilMoveEnv
```

**Comment by duncan-truss** [comment]
This may be leftover from @ronaktruss 's PR but the prime workflow failed for me because I think the prime list_moves task has been removed

```
[2022-01-12 14:19:13,433] Duncans-MacBook-Pro.local/ERROR/locust.user.task: 'PrimeWorkflowTasks' object has no attribute 'list_moves'
Traceback (most recent call last):
  File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 296, in run
    self.execute_next_task()
  File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 321, in execute_next_task
    self.execute_task(self._task_queue.pop(0))
  File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 431, in execute_task
    task(self.user).run()
  File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 281, in run
    self.on_start()
  File "/Users/duncan/workspace/milmove_load_testing/tasks/prime_hhg_workflow.py", line 41, in on_start
    self.tasks = [self.hhg_move, self.list_moves]
AttributeError: 'PrimeWorkflowTasks' object has no attribute 'list_moves'
```

**Comment by felipe-lee** [comment]
> Is `host` a required flag? I feel like before if it wasn't specified we defaulted to local but I get this error if I omit it
> 
> ```
> ➜  milmove_load_testing git:(MB-10977-fl-update-existing-load-tests-to-use-new-helpers) ✗ locust -f locustfiles/prime.py
> [2022-01-12 14:13:39,337] Duncans-MacBook-Pro.local/INFO/locust.main: Starting web interface at http://0.0.0.0:8089 (accepting connections from all network interfaces)
> [2022-01-12 14:13:39,351] Duncans-MacBook-Pro.local/ERROR/root: Uncaught exception in event handler: 
> Traceback (most recent call last):
>   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/event.py", line 40, in fire
>     handler(**kwargs)
>   File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 48, in on_init
>     raise err
>   File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 42, in on_init
>     milmove_env = MilMoveEnv(value=environment.host)
>   File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 384, in __call__
>     return cls.__new__(cls, value)
>   File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 702, in __new__
>     raise ve_exc
> ValueError: None is not a valid MilMoveEnv
> ```

I think it's always been required? At least, I get that same failure in `master` without specifying a host.

**Comment by felipe-lee** [comment]
> This may be leftover from @ronaktruss 's PR but the prime workflow failed for me because I think the prime list_moves task has been removed
> 
> ```
> [2022-01-12 14:19:13,433] Duncans-MacBook-Pro.local/ERROR/locust.user.task: 'PrimeWorkflowTasks' object has no attribute 'list_moves'
> Traceback (most recent call last):
>   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 296, in run
>     self.execute_next_task()
>   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 321, in execute_next_task
>     self.execute_task(self._task_queue.pop(0))
>   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 431, in execute_task
>     task(self.user).run()
>   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/user/task.py", line 281, in run
>     self.on_start()
>   File "/Users/duncan/workspace/milmove_load_testing/tasks/prime_hhg_workflow.py", line 41, in on_start
>     self.tasks = [self.hhg_move, self.list_moves]
> AttributeError: 'PrimeWorkflowTasks' object has no attribute 'list_moves'
> ```

I removed the references to the `list_moves` function, but now I'm getting other errors haha.

**Comment by duncan-truss** [comment]
> > Is `host` a required flag? I feel like before if it wasn't specified we defaulted to local but I get this error if I omit it
> > ```
> > ➜  milmove_load_testing git:(MB-10977-fl-update-existing-load-tests-to-use-new-helpers) ✗ locust -f locustfiles/prime.py
> > [2022-01-12 14:13:39,337] Duncans-MacBook-Pro.local/INFO/locust.main: Starting web interface at http://0.0.0.0:8089 (accepting connections from all network interfaces)
> > [2022-01-12 14:13:39,351] Duncans-MacBook-Pro.local/ERROR/root: Uncaught exception in event handler: 
> > Traceback (most recent call last):
> >   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/event.py", line 40, in fire
> >     handler(**kwargs)
> >   File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 48, in on_init
> >     raise err
> >   File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 42, in on_init
> >     milmove_env = MilMoveEnv(value=environment.host)
> >   File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 384, in __call__
> >     return cls.__new__(cls, value)
> >   File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 702, in __new__
> >     raise ve_exc
> > ValueError: None is not a valid MilMoveEnv
> > ```
> 
> I think it's always been required? At least, I get that same failure in `master` without specifying a host.

Yea I think it's an optional arg because locust doesn't abort from starting but our config doesn't work without it.  I know our Makefile specifies for the common commands it could have been required in a previous version too.

**Comment by felipe-lee** [comment]
> > > Is `host` a required flag? I feel like before if it wasn't specified we defaulted to local but I get this error if I omit it
> > > ```
> > > ➜  milmove_load_testing git:(MB-10977-fl-update-existing-load-tests-to-use-new-helpers) ✗ locust -f locustfiles/prime.py
> > > [2022-01-12 14:13:39,337] Duncans-MacBook-Pro.local/INFO/locust.main: Starting web interface at http://0.0.0.0:8089 (accepting connections from all network interfaces)
> > > [2022-01-12 14:13:39,351] Duncans-MacBook-Pro.local/ERROR/root: Uncaught exception in event handler: 
> > > Traceback (most recent call last):
> > >   File "/Users/duncan/.local/share/virtualenvs/milmove_load_testing-7o6dLg4P/lib/python3.9/site-packages/locust/event.py", line 40, in fire
> > >     handler(**kwargs)
> > >   File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 48, in on_init
> > >     raise err
> > >   File "/Users/duncan/workspace/milmove_load_testing/locustfiles/prime.py", line 42, in on_init
> > >     milmove_env = MilMoveEnv(value=environment.host)
> > >   File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 384, in __call__
> > >     return cls.__new__(cls, value)
> > >   File "/Users/duncan/.pyenv/versions/3.9.6/lib/python3.9/enum.py", line 702, in __new__
> > >     raise ve_exc
> > > ValueError: None is not a valid MilMoveEnv
> > > ```
> > 
> > 
> > I think it's always been required? At least, I get that same failure in `master` without specifying a host.
> 
> Yea I think it's an optional arg because locust doesn't abort from starting but our config doesn't work without it. I know our Makefile specifies for the common commands it could have been required in a previous version too.

Yeah, I think it might default to localhost or something like that. Honestly we aren't really using it as locust means for `host` to be used anyways, so we could do something else if we wanted to. Like base it on an env var, or add a custom argument to our locustfiles for this.

**Review by duncan-truss — APPROVED** [review]
Tested this out locally and it looks good to me.  I did want to point out our ballooning requests to the prime getMove endpoint seems to be 10x as many requests as the others.

Screen Shot 2022-01-12 at 3 26 19 PM

**Comment by felipe-lee** [comment]
> Tested this out locally and it looks good to me. I did want to point out our ballooning requests to the prime getMove endpoint seems to be 10x as many requests as the others.
> 
> Screen Shot 2022-01-12 at 3 26 19 PM

Yeah I think this is likely because of a few changes over the past several prs. One made every function in the prime workflow that gets the current move re-retrieve the move and another made the task that retrieves a move twice as likely to happen as other tasks. There were also a few load tests that did this on their own before doing what they were attempting to actually test. I think in the future (once storage is re-thought out), we might still have really high numbers of hits to that endpoint, but I think we might have a more balanced number (since I think we could likely take off that 2x probably of it happening).

**Pull request merged**